### ISSUE TITLE
arm64 images are using 64bit images and binaries

### ISSUE THREAD

**Issue by ags131** [rook/rook#1872]
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**
arm64 images download, then contain x86_64 binaries

**Expected behavior:**
arm64 binaries

**How to reproduce it (minimal and precise):**

Pull on an arm64 device, image will refuse to run. Extracting a binary and running through `file` shows it to be x86_64.
I checked the binary format with the following commands:
```shell
$ docker run --rm -it -v $PWD/tmp:/app busybox cp /bin/busybox /app/
$ docker run --rm -it -v $PWD/tmp:/app --entrypoint /app/busybox rook/ceph:master cp /bin/sh /app/
$ file tmp/sh
tmp/sh: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, for GNU/Linux 2.6.32, BuildID[sha1]=765c505bb8a234fcd64ede405fa7fcb25734f06a, stripped
```

 https://i.imgur.com/xEC3Ryx.png

**Environment**:
* OS (e.g. from /etc/os-release): HypriotOS 64bit build (Debian 9 Stretch)
* Kernel (e.g. `uname -a`): Linux pi-cluster-2 4.14.37-hypriotos-v8 #1 SMP PREEMPT Sun Apr 29 17:26:16 UTC 2018 aarch64 GNU/Linux
* Cloud provider or hardware configuration: Raspberry Pi 3B+
* Rook version (use `rook version` inside of a Rook Pod): master
* Kubernetes version (use `kubectl version`): 1.10.5
* Kubernetes cluster type (e.g. Tectonic, GKE, OpenShift): RKE
* Storage backend status (e.g. for Ceph use `ceph health` in the [Rook Ceph toolbox](https://rook.io/docs/Rook/master/toolbox.html)): N/A

**Comment by bassam** [comment]
@BlaineEXE @leseb do you have time to take a look here? It looks like the base ceph image is the wrong arch.

**Comment by leseb** [comment]
This is using `ceph/daemon-base:v3.0.5-stable-3.0-luminous-centos-7` right?

**Comment by travisn** [comment]
@leseb yes that's correct. See https://github.com/rook/rook/blob/master/images/Makefile#L19

**Comment by leseb** [comment]
Hum I have a different behaviour when pulling `ceph/daemon-base:v3.0.5-stable-3.0-luminous-centos-7`, see:

```
jenkins-build@omani010:~$ docker run -ti ceph/daemon-base:v3.0.5-stable-3.0-luminous-centos-7 bash
Unable to find image 'ceph/daemon-base:v3.0.5-stable-3.0-luminous-centos-7' locally
v3.0.5-stable-3.0-luminous-centos-7: Pulling from ceph/daemon-base
1e9dd4c85ba2: Already exists
9bfff0e3febd: Pull complete
Digest: sha256:9a0879edee0730e99de3ff48672804c9373503e67bfe5fcf8e3d1b1571c92a4a
Status: Downloaded newer image for ceph/daemon-base:v3.0.5-stable-3.0-luminous-centos-7

jenkins-build@omani010:~$ docker run --rm -it -v $PWD/tmp:/app busybox cp /bin/busybox /app/
Unable to find image 'busybox:latest' locally
latest: Pulling from library/busybox
778a4afaeeb0: Pull complete
Digest: sha256:141c253bc4c3fd0a201d32dc1f493bcf3fff003b6df416dea4f41046e0f37d47
Status: Downloaded newer image for busybox:latest

jenkins-build@omani010:~$ docker run --rm -it -v $PWD/tmp:/app --entrypoint /app/busybox bc2951f1d7a9 cp /bin/sh /app/

jenkins-build@omani010:~$ file tmp/sh
tmp/sh: ELF 64-bit LSB executable, ARM aarch64, version 1 (SYSV), dynamically linked, interpreter /lib/ld-linux-aarch64.so.1, for GNU/Linux 3.7.0, BuildID[sha1]=fd4e6b827c591f90d540582185ef02c5667c0f04, stripped
```

Why don't you simply use `ceph/daemon-base:v3.0.5-stable-3.0-luminous-centos-7`? The Docker Hub is set up with a manifest that contains both arm64 and amd64 archs.
Thanks!

**Comment by travisn** [comment]
@leseb agreed that should work, I see why the build machine would always pick up amd64 for the base image. I'll try this, thanks.

**Comment by arm64b** [comment]
> Why don't you simply use ceph/daemon-base:v3.0.5-stable-3.0-luminous-centos-7? The Docker Hub is set up with a manifest that contains both arm64 and amd64 archs.

Although the `ceph/daemon-base` is multi-arch already(support both `arm64` and `amd64`), but docker doesn't support to pull a different arch image than the host running the `docker pull`, IOW, `docker pull` doesn't support the `--platform` flag yet, so if you run `docker pull ceph/daemon-base:v3.0.5-stable-3.0-luminous-centos-7` on `amd64`, it always to pick up the arch matching the host arch, the `amd64` in this case.

**Comment by leseb** [comment]
@arm64b do you expect a different behavior?

**Comment by bassam** [comment]
@arm64b we're not relying on docker to pick the arch of the base ceph image. We're explicitly passing in `PLATFORM_ARCH`. With #1887 the issues of picking images is fixed, but unfortunately in uncovered another issue.

**Comment by travisn** [comment]
@leseb @ErwanAliasr1 could you look at this issue with the arm64 image? With the fix currently jn #1887, the issue now is that `yum install` always fails for the arm image. The command can be seen on [this line](https://github.com/rook/rook/blob/master/images/ceph/Dockerfile#L17) for the rook ceph image and [this line](https://github.com/rook/rook/blob/master/images/ceph-toolbox/Dockerfile.base#L22) for the rook tools image. The error message can be seen in the PR [here](https://github.com/rook/rook/pull/1887#issuecomment-404962803).

For a simple repro, only the base ceph image is needed and rook is not involved.
```
docker run -it  ceph/daemon-base:v3.0.5-stable-3.0-luminous-centos-7-aarch64
yum install net-tools
```

What are we missing so that the `yum install` command will succeed inside this image? Thanks!

**Comment by leseb** [comment]
I cannot reproduce this, see:

```
jenkins-build@omani010:~$ arch                                                                                                                                                                                                                                                                                       [15/1863]
aarch64
jenkins-build@omani010:~$ docker run -it  ceph/daemon-base:v3.0.5-stable-3.0-luminous-centos-7-aarch64
bash: warning: setlocale: LC_CTYPE: cannot change locale (en_US.UTF-8): No such file or directory
bash: warning: setlocale: LC_COLLATE: cannot change locale (en_US.UTF-8): No such file or directory
bash: warning: setlocale: LC_MESSAGES: cannot change locale (en_US.UTF-8): No such file or directory
bash: warning: setlocale: LC_NUMERIC: cannot change locale (en_US.UTF-8): No such file or directory
bash: warning: setlocale: LC_TIME: cannot change locale (en_US.UTF-8): No such file or directory
[root@54f9a512563a /]# yum --assumeyes install         net-tools         nmap-ncat &&     yum clean all && rm -rf /tmp/* /var/tmp/*
Failed to set locale, defaulting to C
Loaded plugins: fastestmirror, ovl
Determining fastest mirrors
epel/aarch64/metalink                                                                                                                                                                                                                                                                                  |  18 kB  00:00:00
 * epel: mirror.us.leaseweb.net
Ceph                                                                                                                                                                                                                                                                                                   | 2.9 kB  00:00:00
Ceph-noarch                                                                                                                                                                                                                                                                                            | 2.9 kB  00:00:00
base                                                                                                                                                                                                                                                                                                   | 3.6 kB  00:00:00
ceph-source                                                                                                                                                                                                                                                                                            | 2.9 kB  00:00:00
epel                                                                                                                                                                                                                                                                                                   | 3.2 kB  00:00:00
extras                                                                                                                                                                                                                                                                                                 | 2.9 kB  00:00:00
updates                                                                                                                                                                                                                                                                                                | 2.9 kB  00:00:00
(1/10): Ceph-noarch/primary_db                                                                                                                                                                                                                                                                         | 6.7 kB  00:00:00
(2/10): ceph-source/primary_db                                                                                                                                                                                                                                                                         |  23 kB  00:00:00
(3/10): base/7/aarch64/group_gz                                                                                                                                                                                                                                                                        | 140 kB  00:00:00
(4/10): epel/aarch64/group_gz                                                                                                                                                                                                                                                                          |  88 kB  00:00:00
(5/10): Ceph/aarch64/primary_db                                                                                                                                                                                                                                                                        | 307 kB  00:00:00
(6/10): epel/aarch64/updateinfo                                                                                                                                                                                                                                                                        | 925 kB  00:00:00
(7/10): extras/7/aarch64/primary_db                                                                                                                                                                                                                                                                    | 174 kB  00:00:00
(8/10): base/7/aarch64/primary_db                                                                                                                                                                                                                                                                      | 4.8 MB  00:00:01
(9/10): updates/7/aarch64/primary_db                                                                                                                                                                                                                                                                   | 1.2 MB  00:00:00
(10/10): epel/aarch64/primary                                                                                                                                                                                                                                                                          | 3.4 MB  00:00:01
epel                                                                                                                                                                                                                                                                                                              12202/12202
Resolving Dependencies
--> Running transaction check
---> Package net-tools.aarch64 0:2.0-0.22.20131004git.el7 will be installed
---> Package nmap-ncat.aarch64 2:6.40-13.el7 will be installed
--> Processing Dependency: libpcap.so.1()(64bit) for package: 2:nmap-ncat-6.40-13.el7.aarch64
--> Running transaction check
---> Package libpcap.aarch64 14:1.5.3-11.el7 will be installed
--> Finished Dependency Resolution

Dependencies Resolved

==============================================================================================================================================================================================================================================================================================================================
 Package                                                                    Arch                                                                     Version                                                                                      Repository                                                             Size
==============================================================================================================================================================================================================================================================================================================================
Installing:
 net-tools                                                                  aarch64                                                                  2.0-0.22.20131004git.el7                                                                     base                                                                  306 k
 nmap-ncat                                                                  aarch64                                                                  2:6.40-13.el7                                                                                base                                                                  198 k
Installing for dependencies:
 libpcap                                                                    aarch64                                                                  14:1.5.3-11.el7                                                                              base                                                                  132 k

Transaction Summary
==============================================================================================================================================================================================================================================================================================================================
Install  2 Packages (+1 Dependent package)

Total download size: 637 k
Installed size: 2.1 M
Downloading packages:
(1/3): libpcap-1.5.3-11.el7.aarch64.rpm                                                                                                                                                                                                                                                                | 132 kB  00:00:00
(2/3): net-tools-2.0-0.22.20131004git.el7.aarch64.rpm                                                                                                                                                                                                                                                  | 306 kB  00:00:00
(3/3): nmap-ncat-6.40-13.el7.aarch64.rpm                                                                                                                                                                                                                                                               | 198 kB  00:00:00
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Total                                                                                                                                                                                                                                                                                         1.1 MB/s | 637 kB  00:00:00
Running transaction check
Running transaction test
Transaction test succeeded
Running transaction
  Installing : 14:libpcap-1.5.3-11.el7.aarch64                                                                                                                                                                                                                                                                            1/3
  Installing : 2:nmap-ncat-6.40-13.el7.aarch64                                                                                                                                                                                                                                                                            2/3
  Installing : net-tools-2.0-0.22.20131004git.el7.aarch64                                                                                                                                                                                                                                                                 3/3
  Verifying  : net-tools-2.0-0.22.20131004git.el7.aarch64                                                                                                                                                                                                                                                                 1/3
  Verifying  : 2:nmap-ncat-6.40-13.el7.aarch64                                                                                                                                                                                                                                                                            2/3
  Verifying  : 14:libpcap-1.5.3-11.el7.aarch64                                                                                                                                                                                                                                                                            3/3

Installed:
  net-tools.aarch64 0:2.0-0.22.20131004git.el7                                                                                                                         nmap-ncat.aarch64 2:6.40-13.el7

Dependency Installed:
  libpcap.aarch64 14:1.5.3-11.el7

Complete!
Failed to set locale, defaulting to C
Loaded plugins: fastestmirror, ovl
Cleaning repos: Ceph Ceph-noarch base ceph-source epel extras updates
Cleaning up everything
Maybe you want: rm -rf /var/cache/yum, to also free up space taken by orphaned data from disabled or removed repos
Cleaning up list of fastest mirrors
[root@54f9a512563a /]#

**Comment by travisn** [comment]
@bassam we are the only ones that have seen the failure from our macs. Since @BlaineEXE and @leseb have both confirmed the arm64 build succeeds and the image is correct, and our jenkins build has succeeded, I would recommend we merge #1887. To unblock the 0.8 release, we can open a separate issue for the cross builds on mac. Just to confirm the jenkins build is consistent, I just triggered a new build.

**Comment by bassam** [comment]
yes just confirmed that this does not happen on linux -- just Mac with docker for Mac. We still need to fix this but it does not block 0.8

**Comment by jbw976** [comment]
The first image build (`rook/ceph-arm64:v0.7.0-294.gd537df8`) from master looks OK with the fix:
```
> file sh
sh: ELF 64-bit LSB executable, ARM aarch64, version 1 (SYSV), dynamically linked, interpreter /lib/ld-linux-aarch64.so.1, for GNU/Linux 3.7.0, BuildID[sha1]=fd4e6b827c591f90d540582185ef02c5667c0f04, stripped
```
